src/serializers/type_serializers/union.rs

```diff
@@ -198,24 +434,23 @@ impl TypeSerializer for UnionSerializer {
     }
 }
 
-pub struct TaggedUnionBuilder;
-
-impl BuildSerializer for TaggedUnionBuilder {
-    const EXPECTED_TYPE: &'static str = "tagged-union";
-
-    fn build(
-        schema: &Bound<'_, PyDict>,
-        config: Option<&Bound<'_, PyDict>>,
-        definitions: &mut DefinitionsBuilder<CombinedSerializer>,
-    ) -> PyResult<CombinedSerializer> {
-        let schema_choices: Bound<'_, PyDict> = schema.get_as_req(intern!(schema.py(), "choices"))?;
-        let mut choices: Vec<CombinedSerializer> = Vec::with_capacity(schema_choices.len());
-
-        for (_, value) in schema_choices {
-            if let Ok(choice_schema) = value.downcast::<PyDict>() {
-                choices.push(CombinedSerializer::build(choice_schema, config, definitions)?);
-            }
+impl TaggedUnionSerializer {
+    fn get_discriminator_value(&self, value: &Bound<'_, PyAny>, extra: &Extra) -> Option<Py<PyAny>> {
+        let py = value.py();
+        let discriminator_value = match &self.discriminator {
+            Discriminator::LookupKey(lookup_key) => match lookup_key {
+                LookupKey::Simple { py_key, .. } => value.getattr(py_key).ok().map(|obj| obj.to_object(py)),
+                _ => None,
+            },
+            Discriminator::Function(func) => func.call1(py, (value,)).ok(),
+        };
+        if discriminator_value.is_none() {
+            extra.warnings.custom_warning(
+                format!(
+                    "Failed to get discriminator value for tagged union serialization for {value} - defaulting to left to right union serialization."
```

Comment:
We probably need to use the `truncate_input` logic here to avoid crazy error messages cf #1377

Reply:
Done!